addons/product/models/product_template.py

```diff
@@ -178,7 +178,7 @@ def _compute_has_configurable_attributes(self):
 
     @api.depends('product_variant_ids')
     def _compute_product_variant_id(self):
-        for p in self:
+        for p in self.with_context(active_test=False):
             p.product_variant_id = p.product_variant_ids[:1].id
```

Comment:
I fear that you might assign an archived variant as main variant of an active template.
We have to make sure we only fall back on archived variants if there are no active ones.

Reply:
Oof, good catch. I'll do a little tinkering.

Reply:
Committed a potential fix:

```diff
-        for p in self.with_context(active_test=False):
-            p.product_variant_id = p.product_variant_ids[:1].id
+        for p in self:
+            p.product_variant_id = p.with_context(active_test=p.active).product_variant_ids[:1].id
```

The assumption here is that the active_test should be ON when the template is itself active.

Reversely, it doesn't make sense to sort active variants before inactive variants if the template is inactive, because the variants should ALL be inactive, anyway.

I did not write a test for this hyper-specific case. Please let me know if I should.

Reply:
Squashed the fixup commit, then :) Thanks!

Reply:
Thought about it more, and even though I don't think it's gonna break anything, I believe it's conceptually wrong to have the first inactive variant as `product_variant_id` of a template.  As mentioned in my comment, it could be an invalid variant that was archived way before the template was archived, and you might expect it to be a 'valid' configuration when in fact it isn't at all.

* Create a template with 2 attributes with 2 values, sell the different variants
* remove some choices, some of the variants are not valid anymore and archived automatically, 
* archive the template and trigger any logic on it, the `product_variant_id` might have been set to one of the archived, which is not representative of the current state of the template.

You probably should use another way to get a variant of the template when it's archived (maybe something like `_get_variant_for_combination(_get_first_possible_combination())`.  Not sure it'll work fine, but I don't think that setting the first unarchived variant as 'first variant' on an archived template is meaningful, and it might be confusing to developers if it doesn't have the same value than when the template wasn't archived :thinking:

Reply:
Hi! Thanks for the thoughts.

> archive the template and trigger any logic on it, the product_variant_id might have been set to one of the archived, which is not representative of the current state of the template.

I don't think this is correct. Or rather, this is entirely expected. Following the steps:

1. Create a template which has 2 product variants (Conference Chair demo product).
2. Archive one variant.
3. Archive the template.

At or after step 3, step 2 is _lost_. If you subsequently unarchive the template, _all_ variants are also unarchived. If this shouldn't be the case, then we are talking about a bug/missing feature in Odoo. If this isn't a bug or missing feature, though, then my PR is in line with the current behaviour. Although you could argue that it's the unarchiving of the template which introduces the behaviour, not its archiving.

Reply:
For your steps, there won't be any problem indeed, because all the variants of the template (archived or not) are valid variants.  This won't be the case if you remove an attribute, a value, or if you add new exclusions.  In all those situations, if the variant was used before, it'll be archived (but won't be unarchived even if you archive and unarchive the template).  And in those situations, you might get an invalid `product_variant_id`, because there is no check whether the first variant found is valid or not (and I don't think we want to handle that here).

Reply:
Ah! I understand better now. I thought you were directly archiving a variant instead of deleting an attribute value.

In that case I don't know what to do :( Engineering a solution that works for that scenario is more involved than I feel comfortable in re writing PRs against odoo/odoo.

Reply:
Well, I think you should handle the case where the template has no `product_variant_id` in your code instead of modifying the existing logic here :confused: